### ISSUE TITLE
Download design issue

### ISSUE THREAD

**Issue by jonas-jasas** [immich-app/immich#14725]
### The bug

The Immich web interface first downloads images, videos, or albums to memory before saving them to the drive. This approach makes it impossible to download even relatively small pieces of data on mobile devices and becomes an issue on desktops when dealing with larger albums.

### The OS that Immich Server is running on

Any

### Version of Immich Server

Any

### Version of Immich Mobile App

Any

### Platform with the issue

- [ ] Server
- [X] Web
- [ ] Mobile

### Your docker-compose.yml content

```YAML
Design issue
```

### Your .env content

```Shell
Design issue
```

### Reproduction steps

1. Download larger album

**Comment by akostadinov** [comment]
Just FYI, a related issue is that in incognito mode, downloading large files fails.

For example a 2.6GB file outright fails to be downloaded in a Chrome incognito window. And in a Firefox 131 private window it downloads without error but resulting file is a corrupted one of size ~2GB (which is worse because one doesn't expect corrupted files without error).

**Comment by jonas-jasas** [comment]
By fixing this design issue all other issues will be solved automatically.
Looking forward for this issue to be fixed.

**Comment by steve4835** [comment]
Just my 2 cents.  While this is absolutely a valid issue that should be fixed, I will add that by downloading directly to disk instead of saving to memory first, it would not be able (as far as I know) to show the download progress in the Immich UI as it does now.  Maybe add a tweak that checks the file size before downloading and if it is over a certain amount, download it directly to disk, but otherwise keep the current behavior.

**Comment by akostadinov** [comment]
Your browser shows you progress if you care about it. I see no value in another popup. On upload it is a different story. There the progress is important together with the end result.Because these are not reported by the browser.

**Comment by steve4835** [comment]
I don't disagree with that.  It looks nicer when it's integrated with the web UI though.

**Comment by mertalev** [comment]
I agree the best solution would be to leave download progress to the browser. There's no practical reason to require the entire download to fit in RAM.

**Comment by klaus1k** [comment]
Does #16600 help with this issue or is this unrelated?

**Comment by akostadinov** [comment]
idk, need to try it with a new build

**Comment by bo0tzz** [comment]
I doubt it, that's a serverside only change.

**Comment by YapWC** [comment]
I can verify that the problem still exists on v1.129.0 hopefully it gets resolved soon.

**Comment by dragonis41** [comment]
I can confirm that the issue is still present on v1.129.3

If I download an album in incognito mode it fails without more information but **not** in a normal tab.
The generated zip file is estimated to be 3.6GB but it fails at ~54% on a Chrome browser on a Linux computer and at ~1% on a mobile device (still in incognito mode but not in a normal tab)
And in the mobile app, it takes a really long time and stops responding after filled the app cache.

```
[Nest] 17  - 03/27/2025, 5:44:49 PM   ERROR [Api:StreamableFile] Error [ERR_STREAM_PREMATURE_CLOSE]: Premature close
    at ServerResponse.<anonymous> (node:internal/streams/pipeline:415:29)
    at ServerResponse.emit (node:events:530:35)
    at emitCloseNT (node:_http_server:1033:10)
    at Socket.onServerResponseClose (node:_http_server:284:5)
    at Socket.emit (node:events:530:35)
    at TCP.<anonymous> (node:net:351:12)
    at TCP.callbackTrampoline (node:internal/async_hooks:130:17) {
  code: 'ERR_STREAM_PREMATURE_CLOSE'
}
```

I am on the version v1.129.3 and behind a Nginx reverse proxy

**Comment by awnumar** [comment]
One other solution would be for the *server* to compress the files to memory/disk on the server side and send a single blob to the client. That's what Google photos and Google drive does.

**Comment by jonas-jasas** [comment]
It is done by zipping/taring file stream on the fly (HTTP response) without memory/disk.

**Comment by ROODAY** [comment]
Dropping this conversation from Discord: https://discord.com/channels/979116623879368755/1359579903112249424

TL:DR it's seems like it's up to the browser implementation on whether it's in memory or swap or somewhere else?

**Comment by zugao** [comment]
I have the same issues using iPhone. I try to download a 1.6GB of photos via Safari/Chrome and at some point the download frizzes, there are no error logs in nginx or immich containers. The same links works if i download from desktop mac via Chrome.

**Comment by cjvirtucio87** [comment]
this was a blocker for me because I needed to create separate backups (I run immich in kubernetes and have a backup solution, but need to manually download the photos to create a separate backup for 3-2-1). thankfully I found [immich-ruby](https://github.com/chengguangnan/immich-ruby) somewhere. I also created a [fork](https://github.com/cjvirtucio87/immich-python) and re-wrote it in python in case anyone prefers that.

**Comment by artichoked1** [comment]
This seems like a pretty major issue as it makes the download function pretty much unusable on albums bigger than just a handful of images. Not sure why this non-standard design was chosen, apart from allowing a widget to display the progress in the ui? I do not have the skills to contribute and fix this myself, but I am wondering if this is planned to be fixed or not?

**Comment by mertalev** [comment]
We'd like it to be fixed, but we're focused on other tasks at the moment and can't promise when we'll get to this one. A PR to address multi-file downloads would be welcome though.

**Comment by jrasm91** [comment]
There are a lot of implications to this, which I've already discussed at length with other users.

## Current Implementation

1. the client requests download information. 

It'll say "I want to download these 5 assets" or "I want to download this album", etc. The server responds with information like "you want to download these 5 assets as one zip and it'll be this big" or "you want to download this album as 10 zip files, the first zip has these assets and will be this big, and the second zip has these assets and will be this big, etc.". 

2. the client makes the actual download request to the server sending the specific assets. Because it knows the estimated size, it can show a progress indicator. 

## Reasons
- Native downloads require a GET request, and those cannot exceed some maximum length. With the current architecture, we potentially send thousands and thousands of UUIDs, so we will quickly pass that limit, which is why we use a POST with a request body
- The downloads need to be deterministic. If you say "give me the first archive for album 1", "give me the second archive for album 1",  assets may be added or removed to the album while the download is happening, and that will break/affect the offset/pagination you use to determine where archives start and end (we used to do this).
- Native download progress requires a content-length, which we don't know. If we send the wrong value the http request would fail. We get an estimated size from the database, which is good enough for a progress indicator, but not good enough for the content-length header.
- Creating a duplicate zip on disk is slower (read, write, read vs just read), more complicated (needs to have access controls, cleanup, recovery, error handling, and the "link" to the file and its status needs to be communicated asynchronously), and takes up more space on the server (might not be able to download the whole library because the server runs out of space.

## Recommend solution
Personally, if users really want to use the native download feature, I would recommend the following:
- create a table  (`download_requests`) with columns that include at least a `token` field, and an `assetIds` field.
- add an endpoint to create a download request with a list of assetIds, which returns a token or url.
- Use the token/url to start a native browser download (GET `/api/downloads/:token`)

I believe this wouldn't be too complicated and would solve every issue except that you would have no download progress. As a potential workaround for download progress, the download panel could show a link instead of the progress indicator, which could tell you the filename and size( "album_2015...+1.zip (4GiB)"), so you would at least know about how big it is and figure out the progress on your own.

**Comment by mertalev** [comment]
Why couldn't you get download progress with a disk-based solution? Thd client could do a POST request to create a download request and receive a token, at which point the server starts preparing the zip file. When the server then receives the GET request containing the download token, it could wait until the zip file is prepared before responding, then show a clear content length as it starts streaming the file.

**Comment by jonas-jasas** [comment]
If a simple POST request and zip streaming were implemented, it would mean an almost 100% increase in functionality with minimum effort. Additionally, it would ensure 100% compatibility with all HTTP clients. Nobody needs to see a download progress bar for a feature that barely works.

Temporary store UUIDs to Redis when Fetch POST request is made. Return bundle UUID and then change the location (GET request) to download endpoint providing that bundle UUID. In response stream zip with the all files that are in bundle.

**Comment by akostadinov** [comment]
The problem with current implementation is that browsers don't directly save the files but uses some kind of limited cache memory. The bigger the data is, the bigger the chance for a problem is. And one wants a progress bar mostly when download is big.

My suggestion is to use zip streaming and no download indicator, but present a flash message to the user about the estimated download size. So the user will see a flash with "download size 3GB" for example and see in browser's native download info window no progress bar but can see how much was already been downloaded. I think this will be absolutely enough.

P.S. I understand that browsers don't technically have to behave in that way but since they do, arguing with reality is not productive.

**Comment by awnumar** [comment]
> Native download progress requires a content-length, which we don't know.

If the server is already going through the trouble of setting up a table to store the file infos, can it not compute the file-sizes too?

The HTTP layer is already possibly doing gzip compression, and most images and videos will already be compressed so what's the point of wrapping things in another zip file that takes time and space to create? If the zip has compression turned off then is the size deterministic enough to use native browser downloads?

**Comment by jonas-jasas** [comment]
[`Transfer-Encoding: chunked`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Transfer-Encoding#chunked) is your friend.
Regarding temporary asset UUIDs storage: user session storage can be used to simplify even more.

**Comment by jrasm91** [comment]
I feel like nobody is actually adding any value to what I said. All your comments basically say "this is easy, do it this way".

**Comment by AndrewHows** [comment]
Just as an FYI for people stymied by this bug, and can't wait around for it to be fixed...if you setup a storage template that includes the album name, and run the storage template migration job, you can find and retrieve the files from the library on-disk, via the usual file manager for your OS.

**Comment by MrModest** [comment]
Just for the sake of not losing a potentially helpful piece of information:

A Redditor refuses to create a PR, but shares his digging results and potential solution:

https://www.reddit.com/r/selfhosted/s/vWCo4IftH5

> The function `downloadRequest` is the base that other functions seems to be using: https://github.com/immich-app/immich/blob/main/web/src/lib/utils.ts#L107
> 
> It resolves the promise when the entire HTTP request is finished, forcing it to be buffered in memory. This is invoked by `downloadArchive` in `web/src/lib/utils/asset-utils.ts`. `downloadArchive` really only exists to put in GUI updates to the little download manager thingy, and then finally do the same thing as `downloadUrl` but with the pre-prepared download data.
> 
> This hooks into the GUI via the file `actions/DownloadAction.svelte`, where it picks between single file download (that uses the web browser's built-in download manager via `downloadUrl`), and the `downloadArchive` function.
> 
> OK, what do? Well, I think it's better if we just use the web browsers' DL manager instead, so I'd simplify the `downloadArchive` loop to do `downloadUrl` for each single asset, instead of downloading the blob. That'll skip the download GUI entirely, so it won't look the same. As I said, this will remove support for the DL manager thingy, and just use the built-in one.
> 
> The "best" thing to do is probably to use the downloads API, but that doesn't have that much support. https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/downloads
> 
> Anyway, the code change is very simple. The rest is to get people to feel like this is the good thing to do.
> 
> Feel free to copy+paste this to the Github issue, CBA personally.
> 
> Edit, fuck it, this should be what it should look like. This'll create one download per archive:
> ```ts
> export const downloadArchive = async (fileName: string, options: Omit<DownloadInfoDto, 'archiveSize'>) => { 
>     const $preferences = get<UserPreferencesResponseDto | undefined>(preferences); 
>     const dto = { ...options, archiveSize: $preferences?.download.archiveSize };
> 
>     const [error, downloadInfo] = await withError(() => getDownloadInfo({ ...authManager.params, downloadInfoDto: dto }));
>     if (error) {
>        const $t = get(t);
>        handleError(error, $t('errors.unable_to_download_files'));
>        return;
>     }
> 
>     if (!downloadInfo) {
>        return;
>     }
> 
>     for (let index = 0; index < downloadInfo.archives.length; index++) {
>        const archive = downloadInfo.archives[index];
>        const suffix = downloadInfo.archives.length > 1 ? `+${index + 1}` : '';
>        const archiveName = fileName.replace('.zip', `${suffix}-${DateTime.now().toFormat('yyyyLLdd_HHmmss')}.zip`);
>        const queryParams = asQueryString(authManager.params);
>        const url = getBaseUrl() + '/download/archive' + (queryParams ? `?${queryParams}` : '';
> 
>        let downloadKey = `${archiveName} `;
>        if (downloadInfo.archives.length > 1) {
>            downloadKey = `${archiveName} (${index + 1}/${downloadInfo.archives.length})`;
>        }
> 
>        try {
>            downloadUrl(url, archiveName);
>        } catch (error) {
>            const $t = get(t);
>            handleError(error, $t('errors.unable_to_download_files'));
>            return;
>        }
>     }
> 
> }; 
> ```

**Comment by diogotcorreia** [comment]
> Personally, if users really want to use the native download feature, I would recommend the following:
>
> - create a table (download_requests) with columns that include at least a token field, and an assetIds field.
> - add an endpoint to create a download request with a list of assetIds, which returns a token or url.
> - Use the token/url to start a native browser download (GET /api/downloads/:token)
>
> I believe this wouldn't be too complicated and would solve every issue except that you would have no download progress. As a potential workaround for download progress, the download panel could show a link instead of the progress indicator, which could tell you the filename and size( "album_2015...+1.zip (4GiB)"), so you would at least know about how big it is and figure out the progress on your own.

I agree with this solution by @jrasm91, but some further notes:
- For authentication, do not link the current logged in user to the "token", since the permissions can be revoked (e.g., user stops having access to album, shared linked is deleted, etc). Instead, perform the authentication at download time (as it is currently being done), and just use the "token" as a collection of asset IDs.
- The `download_requests` table needs to be cleared once in a while, I would suggest a "token" lifespan of at least 24h.
- The `/download/info` endpoint would no longer need to return a list of all assetIds, potentially decreasing the response size significantly (good on slow connections!). However, this would be a breaking change, so care needs to be taken here. Perhaps create a new `/download/prepare` endpoint instead for this new functionality.

This would already be enough to have feature parity with Apple's iCloud share links on the web, but I think we can do better and have a proper progress bar and even resumable downloads.

> If the server is already going through the trouble of setting up a table to store the file infos, can it not compute the file-sizes too?
>
> The HTTP layer is already possibly doing gzip compression, and most images and videos will already be compressed so what's the point of wrapping things in another zip file that takes time and space to create? If the zip has compression turned off then is the size deterministic enough to use native browser downloads?

For this reason, I like @awnumar 's solution: most photos/videos are barely compressable, and the `/download/archive` endpoint is already using brotli compression for me anyway. There is no need to compress the files inside the zip, as is already the case for the current implementation ([`{ store: true }`](https://www.archiverjs.com/docs/archiver#zip-entry-properties)): https://github.com/immich-app/immich/blob/8fe54a4de17bd79ae2f9fa8fd4892473d13b0618/server/src/repositories/storage.repository.ts#L85
If we do not compress files inside the zip file, it is possible to calculate the zip file's exact size beforehand, as done in the zipstream-ng project: https://github.com/pR0Ps/zipstream-ng/blob/991d6cb33361d9071948e5b625cebc2ae1d45625/zipstream/ng.py#L553-L565 https://github.com/pR0Ps/zipstream-ng/blob/991d6cb33361d9071948e5b625cebc2ae1d45625/zipstream/ng.py#L1210
It's not as simple as the current size estimation in Immich, but I'd say it is pretty doable with minimal overhead.

This definitely enables a progress bar in the browser, and opens the door to support [resumable downloads](https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/Range_requests) in the future.

> Why couldn't you get download progress with a disk-based solution? Thd client could do a POST request to create a download request and receive a token, at which point the server starts preparing the zip file. When the server then receives the GET request containing the download token, it could wait until the zip file is prepared before responding, then show a clear content length as it starts streaming the file.

@mertalev I disagree with a disk-based solution because it can lead to denial of service by resource exhaustion, or even just not work at all if the server has disk space constraints. This is obvious in the extreme case where a user tries to download an album with hundreds of GBs of assets, and the server might not have that available disk space.

---

I'm busy for the next 2 weeks or so, but wouldn't mind working on this if there is no progress meanwhile.

**Comment by aleqx** [comment]
> I'm busy for the next 2 weeks or so, but wouldn't mind working on this if there is no progress meanwhile.

Wondering if you've got around to it. Thanks in advance.

**Comment by diogotcorreia** [comment]
> Wondering if you've got around to it. Thanks in advance.

@aleqx Hey, sorry I kinda forgot about this, it got buried in my inbox 😅 I don't think I'll be able to look at this this week, but I will hopefully have some time next week 🤞

**Comment by diogotcorreia** [comment]
@aleqx I have started to work on this on https://github.com/immich-app/immich/pull/24121
Very much still a work-in-progress, and obviously cannot promise any ETA, but you can follow that PR for progress.

**Comment by aleqx** [comment]
> [@aleqx](https://github.com/aleqx) I have started to work on this on [#24121](https://github.com/immich-app/immich/pull/24121) Very much still a work-in-progress, and obviously cannot promise any ETA, but you can follow that PR for progress.

Great stuff! Many thanks for letting us know

**Comment by b4stOss** [comment]
Any update on this one? Just encountered this bug and it's basically just impossible to download any album. Quite inconvenient

**Comment by diogotcorreia** [comment]
Hey @b4stOss (and others following this issue)
I have https://github.com/immich-app/immich/pull/24121 open, but unfortunately haven't had much time to finish it.
The backend part is practically done, but I'm struggling with the frontend/UX. If you have ideas for that, please join the discussion on that PR.

**Comment by patienttruth** [comment]
@b4stOss I ran into this issue today with a 150mb download, but it was present even with 20mb. Per a comment in this thread or another I found my primary OS drive (Windows 10) had less than 4gb available. Note, my Downloads folder is on another drive with plenty of space, however I cleared out some old apps and files so I had more room. Now download work normally again. From that experience and reading this thread it appears the browser uses the OS drive for caching, so you run into this error when it's low on space.

**Comment by fredfloydd** [comment]
Looks like the PR to fix this has stalled a bit - any plans to pick it back up @diogotcorreia? If not I'll probably take a look at doing...

**Comment by jonas-jasas** [comment]
I have an idea how to fix this issue and keep progress bar.
When the user clicks to download photos, he must [choose a directory](https://developer.mozilla.org/en-US/docs/Web/API/Window/showDirectoryPicker) where all photos will be saved without zipping. Then using [File System API](https://developer.mozilla.org/en-US/docs/Web/API/File_System_API) save all photos and display progress bar.
If this idea is implemented there will be no need to change server side.

**Comment by fredfloydd** [comment]
I think that might make the UX a bit weird? Having the option to download multiple individual photos without archiving totally makes sense for small numbers, but for large albums of 1000s of photos, it seems pretty impractical to hammer the download manager?

I think there's a place for both, ideas around there being a more proper ui for preparing and managing large downloads makes sense to me

**Comment by jonas-jasas** [comment]
Download manager is not involved when using Fetch API and File System API. There would be no hammering and changed UX.

**Comment by akostadinov** [comment]
I can't remember a situation where I'd like to download a huge ZIP of media. And cannot foresee one. especially for large downloads, it would mean having the disk space and suffering additional time penalty to get to the files. Not to mention the inconvenience.

Just saying... I don't mind having the option to download zips. But it has always appeared to me more of a workaround than actually a feature.

**Comment by patienttruth** [comment]
While not versed on the technical side of implementing this, I'll just say that in general I'd prefer the files to be dumped into a folder, not a zip. I'm downloading to edit and access them for a specific project so it adds a step. That said, I'm a fan of options and can certainly see cases where a zip would be preferred. Perhaps the UI could always ask unless an is override is set.

**Comment by diogotcorreia** [comment]
@fredfloydd I hope to be able to continue working on the PR, yes. Sadly my frontend skills are a bit rusty and I haven't heard back from the Immich team on the questions I asked there, so I might just take the liberty to make some decisions instead.

> I have an idea how to fix this issue and keep progress bar. When the user clicks to download photos, he must [choose a directory](https://developer.mozilla.org/en-US/docs/Web/API/Window/showDirectoryPicker) where all photos will be saved without zipping. Then using [File System API](https://developer.mozilla.org/en-US/docs/Web/API/File_System_API) save all photos and display progress bar. If this idea is implemented there will be no need to change server side.

@jonas-jasas This cannot be used at the moment because it does not work in Firefox nor Safari as per the link you sent (as least `showDirectoryPicker`). I doubt the Immich team wants to maintain different code across browsers. Could be a nice feature to consider in a few years, however.

**Comment by ereynier** [comment]
I reviewed the current `main` implementation, #14725, #17507, and the draft work in #24121. I would be willing to work on a smaller stateless alternative, but I would like to validate the architecture with the maintainers and with @diogotcorreia before starting a public PR.

## Proposed approach

- Keep the existing `POST /download/archive` behavior unchanged for SDK/mobile/third-party clients.
- Extend `POST /download/info` with one short-lived signed ticket per archive.
- Add a web-only endpoint such as `POST /download/archive/browser`.
- Submit that endpoint through a native HTML multipart form, so the browser download manager receives Immich's existing ZIP stream directly (no XHR, Blob, temporary ZIP, database request row, or token in the URL).
- Re-authenticate the request and call the existing `downloadArchive` service so asset permissions are checked again at download time.
- Bind each ticket to:
  - version and audience;
  - approximately five minutes of validity;
  - either the current user ID or the exact shared-link ID;
  - the exact ordered asset ID list;
  - the `edited` value.
- Accept replay during that short lifetime, since the current identity and permissions are checked again.
- Bound the multipart parser and decoded ticket. I am considering an initial cap around 5,000 assets per ticket, with archive splitting also based on asset count.
- For multiple archive parts, show one explicit button per part instead of triggering multiple automatic downloads.
- Redact ticket/token fields from Immich's verbose request-body logging.

This differs from #24121 by avoiding the download-request tables, 24-hour records, cleanup job, and request ID in a GET URL. It also means the response from `/download/info` remains proportional to the number of assets, so the per-ticket asset cap is important.

## Decision needed: signing secret

Current `main` has reusable HS256 signing helpers, but I could not find a persistent general-purpose server secret suitable for these tickets. The workflow JWT secret is process-local and regenerated at startup, while the maintenance secret is purpose-specific and temporary.

Would the preferred convention be:

1. a new general signing secret supplied through configuration;
2. a secret generated once and stored in the existing `system_metadata` table; or
3. another signing mechanism already planned by the project?

Option 2 appears operationally simplest and would not add a new table or any per-download write, but it does add one persistent initialization write and should be explicitly accepted.

Do maintainers consider this stateless POST design acceptable, and would you prefer a new PR or collaborative continuation of #24121?

**Comment by diogotcorreia** [comment]
@ereynier Thank you for you suggestions! I liked the HTML form approach you suggested, but you don't need any ticket/session if you are submitting a POST request, we can just reuse the existing endpoints.

I opened https://github.com/immich-app/immich/pull/30021 for this. It is much much simpler than my original PR, and avoids many of the problems I was having.

It's just missing some cleanup and CSS changes, but other than that should be done. It also needs testing on Safari, just to make sure this will work everywhere.
I might be able to finish these two things tomorrow, but we'll see how it goes.
